telemetry/telemetry.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2017 Microsoft. All rights reserved.
+// MIT License
+
+package telemetry
+
+import (
+	"bytes"
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+)
+
+type xmlDocument struct {
+	XMLName   xml.Name `xml:"Interfaces"`
+	Interface []struct {
+		XMLName    xml.Name `xml:"Interface"`
+		MacAddress string   `xml:"MacAddress,attr"`
+		IsPrimary  bool     `xml:"IsPrimary,attr"`
+
+		IPSubnet []struct {
+			XMLName xml.Name `xml:"IPSubnet"`
+			Prefix  string   `xml:"Prefix,attr"`
+
+			IPAddress []struct {
+				XMLName   xml.Name `xml:"IPAddress"`
+				Address   string   `xml:"Address,attr"`
+				IsPrimary bool     `xml:"IsPrimary,attr"`
+			}
+		}
+	}
+}
+
+// OS Details structure.
+type OSInfo struct {
+	OSType         string
+	OSVersion      string
+	KernelVersion  string
+	OSDistribution string
+}
+
+// System Details structure.
+type SystemInfo struct {
+	MemVMTotal       uint64
+	MemVMFree        uint64
+	MemUsedByProcess uint64
+	DiskVMTotal      uint64
+	DiskVMFree       uint64
+	CPUCount         int
+}
+
+// Interface Details structure.
+type InterfaceInfo struct {
+	InterfaceType         string
+	Subnet                string
+	PrimaryCA             string
+	MAC                   string
+	Name                  string
+	SecondaryCATotalCount int
+	SecondaryCAUsedCount  int
+}
+
+// CNI Bridge Details structure.
+type BridgeInfo struct {
+	NetworkMode string
+	BridgeName  string
+}
+
+type OrchsestratorInfo struct {
+	OrchestratorName    string
+	OrchestratorVersion string
+}
+
+// Azure CNI Telemetry Report structure.
+type Report struct {
+	StartFlag           bool
+	Name                string
+	Version             string
+	OrchestratorDetails *OrchsestratorInfo
+	OSDetails           *OSInfo
+	SystemDetails       *SystemInfo
+	InterfaceDetails    *InterfaceInfo
+	BridgeDetails       *BridgeInfo
+	VnetAddressSpace    []string
+	ErrorMessage        string
+	Context             string
+	SubContext          string
+}
+
+// ReportManager structure.
+type ReportManager struct {
+	HostNetAgentURL string
+	IpamQueryURL    string
+	ReportType      string
+	Report          *Report
+}
+
+// GetReport retrieves orchestrator, system, OS and Interface details and create a report structure.
+func (reportMgr *ReportManager) GetReport(name string, version string) (*Report, error) {
+	var err error
+	report := &Report{}
+
+	report.Name = name
+	report.Version = version
+	report.OrchestratorDetails = GetOrchestratorDetails()
```

Comment:
The original assumption while implementing this function was it will be used by CNM also and in that case this function will fail and we should not return anything. Now I'm thinking of better idea to do that. Will change this function to report error